crates/typst-syntax/src/lexer.rs

```diff
@@ -96,23 +137,46 @@ impl Lexer<'_> {
 }
 
 /// Shared methods with all [`LexMode`].
-impl Lexer<'_> {
-    /// Proceed to the next token and return its [`SyntaxKind`]. Note the
-    /// token could be a [trivia](SyntaxKind::is_trivia).
-    pub fn next(&mut self) -> SyntaxKind {
-        if self.mode == LexMode::Raw {
-            let Some((kind, end)) = self.raw.pop() else {
-                return SyntaxKind::End;
-            };
-            self.s.jump(end);
-            return kind;
+impl<'s> Lexer<'s> {
+    /// Move the lexer forward to return the next token and, in Math/Code mode, lex past
+    /// any trivia tokens, pushing them into `nodes`.
+    pub fn lex_past_trivia(&mut self, nodes: &mut Vec<SyntaxNode>) -> Token {
+        let mut start = self.cursor();
+        let mut kind = self.next();
+        let mut triv = TriviaStart { num: 0, offset: start };
+
+        if self.mode != LexMode::Markup {
+            // Skip past any trivia at the start when in Math/Code mode.
+            while kind.is_trivia() {
+                nodes.push(SyntaxNode::leaf(kind, self.s.from(start)));
+                start = self.cursor();
+                kind = self.next();
+                triv.num += 1;
+            }
         }
+        let prev_trivia = if triv.num != 0 { Some(triv) } else { None };
+        let node = match self.error.take() {
+            Some(error) => {
+                kind = SyntaxKind::Error;
+                SyntaxNode::error(error, self.s.from(start))
+            }
+            None if kind == SyntaxKind::Raw => {
+                SyntaxNode::inner(kind, std::mem::take(&mut self.raw))
+            }
+            None => SyntaxNode::leaf(kind, self.s.from(start)),
+        };
+        Token { node, kind, prev_trivia }
+    }
 
-        self.newline = false;
-        self.error = None;
+    /// Proceed to the next token and return its [`SyntaxKind`] plus a potential error.
+    /// Note the token could be a [trivia](SyntaxKind::is_trivia).
+    pub fn next(&mut self) -> SyntaxKind {
+        assert_eq!(self.raw.len(), 0);
+        assert_eq!(self.error, None);
         let start = self.s.cursor();
```

Comment:
Wouldn't it be cleaner if
- `next` returned a `SyntaxNode`
- The error and raw logic was moved into `next()` (or even further down into the places where it would be created, though then all the parser function would need to return nodes, which might slow things down, not sure)
- The trivia stuff stays in `lex_past_trivia` and moves back into the parser?

This way, the lexer has a clearer public API and is not tied to the `nodes` vector and `Token` struct.